gradle-plugin/src/main/java/com/squareup/anvil/plugin/CheckMixedSourceSet.kt

```diff
@@ -22,44 +23,52 @@ open class CheckMixedSourceSet(
   private val compileTask: KotlinCompile
 ) {
 
-  fun disablePreciseJavaTrackingIfNeeded() {
+  fun shouldDisablePreciseJavaTrackingIfNeeded(): Property<Boolean> {
     val sourceFiles = when {
-      project.isAndroidProject -> getSourceFilesAndroidProject()
-      project.isKotlinJvmProject -> getSourceFilesJavaProject()
+      project.isAndroidProject -> project.getSourceFilesAndroidProject(compileTask)
+      project.isKotlinJvmProject -> project.getSourceFilesJavaProject()
       // Consider Kotlin only project in the future.
       else -> throw GradleException("Only Android and JVM modules are supported for now.")
     }
 
     // If there is Java file, then disable precise Java tracking.
-    val hasJavaFile = sourceFiles.any { it.extension == "java" }
+    val hasJavaFile = project.provider {
+      sourceFiles.any { it.extension == "java" }
+    }
 
     // If Kapt is enabled then it usually generates Java code making Kotlin compilation use a
     // mixed source set.
-    val isKaptApplied = project.plugins.hasPlugin(Kapt3GradleSubplugin::class.java)
-
-    if (hasJavaFile || isKaptApplied) {
-      compileTask.usePreciseJavaTracking = false
+    val isKaptApplied = project.provider {
+      project.plugins.hasPlugin(Kapt3GradleSubplugin::class.java)
     }
+
+    return project.objects.property(Boolean::class.java)
+        .also {
+          it.set(hasJavaFile.get() || isKaptApplied.get())
+        }
   }
 
   @OptIn(ExperimentalStdlibApi::class)
-  private fun getSourceFilesAndroidProject(): Sequence<File> {
-    return project.androidVariants()
-        .findVariantForCompileTask(compileTask)
-        .sourceSets
-        .asSequence()
-        .flatMap { it.javaDirectories.asSequence() }
-        .flatMap { it.walk() }
-        .filter { it.isFile }
-  }
+  private fun Project.getSourceFilesAndroidProject(compileTask: KotlinCompile) =
+    this.files( this.provider {
```

Comment:
I actually missed the provider first. This would be easier to read if you assign the provider to a temporary variable.